phoenix-core/src/main/java/org/apache/phoenix/coprocessor/ChildLinkMetaDataEndpoint.java

```diff
@@ -111,4 +141,245 @@ private PhoenixMetaDataCoprocessorHost getCoprocessorHost() {
+     * An instance of this class is created for each scanner on the table
+     * and used to verify individual rows.
+     */
+    public class ChildLinkMetaDataScanner extends BaseRegionScanner {
+
+        private RegionScanner scanner;
+        private Scan scan;
+        private RegionCoprocessorEnvironment env;
+        private Scan sysCatScan = null;
+        private Scan childLinkScan;
+        private byte[] emptyCF;
+        private byte[] emptyCQ;
+        private Region region;
+        private boolean hasMore;
+        private long pageSizeMs;
+        private long pageSize = Long.MAX_VALUE;
+        private long rowCount = 0;
+        private long maxTimestamp;
+        private long ageThreshold;
+
+        public ChildLinkMetaDataScanner(RegionCoprocessorEnvironment env,
+                                        Scan scan,
+                                        RegionScanner scanner) {
+            super(scanner);
+            this.env = env;
+            this.scan = scan;
+            this.scanner = scanner;
+            region = env.getRegion();
+            emptyCF = scan.getAttribute(EMPTY_COLUMN_FAMILY_NAME);
+            emptyCQ = scan.getAttribute(EMPTY_COLUMN_QUALIFIER_NAME);
+            pageSizeMs = getPageSizeMsForRegionScanner(scan);
+            maxTimestamp = scan.getTimeRange().getMax();
+            ageThreshold = env.getConfiguration().getLong(
+                    QueryServices.CHILD_LINK_ROW_AGE_THRESHOLD_TO_DELETE_MS_ATTRIB,
+                    QueryServicesOptions.DEFAULT_CHILD_LINK_ROW_AGE_THRESHOLD_TO_DELETE_MS);
+        }
+
+        public boolean next(List<Cell> result, boolean raw) throws IOException {
+            try {
+                long startTime = EnvironmentEdgeManager.currentTimeMillis();
+                do {
+                    if (raw) {
+                        hasMore = scanner.nextRaw(result);
+                    } else {
+                        hasMore = scanner.next(result);
+                    }
+                    if (result.isEmpty()) {
+                        break;
+                    }
+                    if (isDummy(result)) {
+                        return true;
+                    }
+                    Cell cell = result.get(0);
+                    if (verifyRowAndRepairIfNecessary(result)) {
+                        break;
+                    }
+                    if (hasMore && (EnvironmentEdgeManager.currentTimeMillis() - startTime) >= pageSizeMs) {
+                        byte[] rowKey = CellUtil.cloneRow(cell);
+                        result.clear();
+                        getDummyResult(rowKey, result);
+                        return true;
+                    }
+                    // skip this row as it is invalid
+                    // if there is no more row, then result will be an empty list
+                } while (hasMore);
+                rowCount++;
+                if (rowCount == pageSize) {
```

Comment:
You need to check if the scan has the page filter. If so, you need to remove it from the scan and then reopen the underlying the region scanner. Please see GlobalIndexChecker.

Reply:
Please add a test to make sure that limit queries return the correct number of rows in the presence of unverified rows.